%% LyX 1.6.7 created this file.  For more info, see http://www.lyx.org/.
%% Do not edit unless you really know what you are doing.
\documentclass[english]{article}
\usepackage{lmodern}
\usepackage[T1]{fontenc}
\usepackage[latin9]{inputenc}
\usepackage[letterpaper]{geometry}
\geometry{verbose,tmargin=2.5cm,bmargin=2.5cm,lmargin=2.5cm,rmargin=2.5cm}
\usepackage{babel}

\usepackage{url}
\usepackage{amssymb}
\usepackage[numbers]{natbib}
\usepackage[unicode=true, pdfusetitle,
 bookmarks=true,bookmarksnumbered=false,bookmarksopen=false,
 breaklinks=false,pdfborder={0 0 1},backref=false,colorlinks=false]
 {hyperref}

\makeatletter

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%% LyX specific LaTeX commands.
%% Because html converters don't know tabularnewline
\providecommand{\tabularnewline}{\\}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%% Textclass specific LaTeX commands.
\usepackage{Sweave}
\newenvironment{wrapsweave}{\par}{\par}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%% User specified LaTeX commands.
%\VignetteIndexEntry{3 - Growth temperature and relative stabilities of proteins}

% note: this is not compatible with multicore; with multicore
% loaded the PDF files of the images are damaged

\makeatother

\begin{document}

\title{Growth temperature and relative stabilities of proteins}


\author{Jeffrey M. Dick}
\maketitle
\begin{abstract}
This vignette demonstrates an analysis of the relative stabilities
of 24 carboxylases from model organisms that grow optimally at different
temperatures. In the chemical thermodynamic model, changes in both
temperature and oxidation potential influence the relative stabilities
of the proteins with an overall progression that fits with the growth
temperatures of the organisms. The calculations supporting this study
are presented using a series of equilibrium activity diagrams constructed
using CHNOSZ. First, equilibrium activity diagrams as a function of
either temperature or $\log a_{\mathrm{H_{2}}}$ (at constant values
of the other variable) show that both variables contribute significantly
to the relative stabilities of the proteins. Second, an equilibrium
predominance diagram is used to formulate an expression for $\log a_{\mathrm{H_{2}}}$
as a function of temperature that crosses the protein stability fields
in order of increasing growth temperature of the organism. Third,
equilibrium activity and chemical species richness diagrams based
on this function show pattern that mirrors the natural growth temperature
of the organism. The final piece of code produces an animated series
of rank-abundance diagrams for the proteins as both temperature and
$\log a_{\mathrm{H_{2}}}$ are increased.
\end{abstract}
\begin{wrapsweave}
<<options,echo=FALSE>>=

options(width=80,continue=" ")

@
\end{wrapsweave}



\section{Introduction}

The motivation for this set of calculations is to explore how differences
in the chemical stabilities of proteins might underlie adaptation
of organisms to a range of environments that vary in temperature and
chemical characteristics. The approach we take is to perform chemical
thermodynamic calculations of the relative stabilities of proteins
from different organisms. These calculations are carried out using
the CHNOSZ package \citep{Dic08} for the R software environment \citep{RDCT}.

A collection of model organisms and proteins is listed in Table \ref{tab:carboxylase}.
The organisms listed in Table \ref{tab:carboxylase} were chosen because
the represent a cross-section of mesophilic and thermophilic organisms.
This list was inspired by recognizing the great diversity of organisms
reported in Yellowstone hot springs (see references in the table),
but it also contains organisms from soils, ocean and submarine hot
spring environments. Where possible, literature reference that identify
types of organisms found in Yellowstone National Park are listed,
as are references giving the optimal growth temperatures of the model
organisms.

Carboxylases were identified as the model proteins in part because
they are found in a wide range organisms and their sequences are often
available, even for those organisms which have not whole genomes available.
Homologs of the enzyme ribulose biphosphate carboxylase (RuBisCo)
have been shown to be useful in exploring evolutionary relationships
of organisms that have this enzyme \citep{Ray05,ABR+06,WRW+09}. Homologs
of both RuBisCo and acetyl-coenzyme A carboxylase (AcCoACo is the
abbreviation used here) are listed in Table \ref{tab:carboxylase}.
The AcCoACo homologs were included in order to expand the range of
organisms considered here.

For each protein family there are twelve organisms listed in Table
\ref{tab:carboxylase}, in approximate order of increasing optimal
growth temperature. For the purposes of the calculations described
below, the first six proteins in each group are classified as coming
from low-temperature environments, and the last six as coming from
high-temperature environments.

%
\begin{table}
\caption{\label{tab:carboxylase}Model organisms and identities of model carboxylases
used in this study. The protein IDs are those used in the SWISS-PROT/UNIPROT
database \citep{BBA+03}; if the ID is a SWISS-PROT name the accession
number is given in parentheses. Abbreviations: A -- archaea; B --
bacteria; E -- eukarya; $T_{\mathrm{opt}}$ -- optimal growth temperature.}


\begin{tabular}{cllllc}
\hline 
{\footnotesize A/B/E} & {\footnotesize Organism} & {\footnotesize $T_{\mathrm{opt}}$, $^{\circ}$C} & {\footnotesize ID} & {\footnotesize Formula} & {\footnotesize Length}\tabularnewline
\hline
\multicolumn{6}{c}{{\footnotesize Ribulose bisphosphate carboxylase (large subunit except
{*})}}\tabularnewline
{\footnotesize B} & \emph{\footnotesize $\alpha$-proteobacterium }{\footnotesize (}\emph{\footnotesize B.
japonicum}{\footnotesize )} & {\footnotesize 25} & {\footnotesize RBL\_BRAJA (Q9ZI34)} & {\footnotesize $\mathrm{C_{2403}H_{3735}N_{663}O_{702}S_{21}}$} & {\footnotesize 486}\tabularnewline
{\footnotesize B} & {\footnotesize $\beta$- }\emph{\footnotesize proteobacterium }{\footnotesize (}\emph{\footnotesize T.
denitrificans}{\footnotesize ) \citep{FB72}} & {\footnotesize 28-32 \citep{KW00}} & {\footnotesize{} A6YF84\_9PROT ({*})} & {\footnotesize $\mathrm{C_{1061}H_{1670}N_{312}O_{299}S_{11}}$} & {\footnotesize 213}\tabularnewline
{\footnotesize E} & \emph{\footnotesize Bracteacoccus giganteus }{\footnotesize \citep{NC06}} & {\footnotesize --} & {\footnotesize A1E8R4\_9CHLO} & {\footnotesize $\mathrm{C_{1325}H_{2091}N_{383}O_{379}S_{14}}$} & {\footnotesize 269}\tabularnewline
{\footnotesize B} & \emph{\footnotesize Mycobacterium}{\footnotesize{} sp. DSM 3803 \citep{WSP05}} & {\footnotesize --} & {\footnotesize A8C9T6\_9MYCO} & {\footnotesize $\mathrm{C_{2344}H_{3614}N_{640}O_{690}S_{19}}$} & {\footnotesize 476}\tabularnewline
{\footnotesize B} & \emph{\footnotesize Leptospirillum }{\footnotesize sp. Group II UBA
\citep{WSP05}} & {\footnotesize $\sim$37} & {\footnotesize A3EQE1\_9BACT} & {\footnotesize $\mathrm{C_{1901}H_{3060}N_{530}O_{543}S_{23}}$} & {\footnotesize 392}\tabularnewline
{\footnotesize B} & \emph{\footnotesize Cyanobium}{\footnotesize{} sp. PCC 7009 \citep{NC06}} & {\footnotesize $\sim$45} & {\footnotesize A5CKC7\_9CHRO} & {\footnotesize $\mathrm{C_{2345}H_{3606}N_{646}O_{687}S_{19}}$} & {\footnotesize 470}\tabularnewline
{\footnotesize B} & \emph{\footnotesize Cyanobacterium}{\footnotesize{} (}\emph{\footnotesize Synechococcus}{\footnotesize )}\emph{\footnotesize{}
}{\footnotesize \citep{ABR+06}} & {\footnotesize 50 \citep{ABR+06}} & {\footnotesize RBL\_SYNJA (Q2JV67)} & {\footnotesize $\mathrm{C_{2366}H_{3652}N_{650}O_{683}S_{21}}$} & {\footnotesize 474}\tabularnewline
{\footnotesize E} & \emph{\footnotesize Cyanidiales}{\footnotesize{} \citep{FB72,WSP05}} & {\footnotesize 50--55 \citep{CYP+04}} & {\footnotesize Q6JAIO\_9RHOD} & {\footnotesize $\mathrm{C_{2029}H_{3169}N_{551}O_{575}S_{16}}$} & {\footnotesize 405}\tabularnewline
{\footnotesize A} & \emph{\footnotesize Methanococcus jannaschii } & {\footnotesize 85 \citep{JLM+83}} & {\footnotesize RBL\_METJA (Q58632) ({*})} & {\footnotesize $\mathrm{C_{2151}H_{3412}N_{574}O_{627}S_{16}}$} & {\footnotesize 425}\tabularnewline
{\footnotesize A} & \emph{\footnotesize Staphylothermus marinus}{\footnotesize{} \citep{MSA05}} & {\footnotesize 85--92 \citep{Ste96,FSJ+86}} & {\footnotesize A3DND9\_STAMF} & {\footnotesize $\mathrm{C_{2264}H_{3529}N_{603}O_{633}S_{13}}$} & {\footnotesize 443}\tabularnewline
{\footnotesize A} & \emph{\footnotesize Thermofilum pendens }{\footnotesize \citep{MSA05}} & {\footnotesize 85--90 \citep{Ste96}} & {\footnotesize A1RZJ5\_THEPD} & {\footnotesize $\mathrm{C_{2215}H_{3468}N_{598}O_{630}S_{13}}$} & {\footnotesize 443}\tabularnewline
{\footnotesize A} & \emph{\footnotesize Pyrococcus horokoshii }{\footnotesize \citep{MSA05}} & {\footnotesize 98 \citep{GMR+98}} & {\footnotesize RBL\_PYRHO (O58677) ({*})} & {\footnotesize $\mathrm{C_{2179}H_{3409}N_{587}O_{615}S_{18}}$} & {\footnotesize 430}\tabularnewline
\hline 
\multicolumn{6}{c}{{\footnotesize Acetyl-coenzyme A carboxylase ({*} carboxyl transferase
subunit alpha)}}\tabularnewline
{\footnotesize B} & {\footnotesize $\gamma$-}\emph{\footnotesize proteobacterium}{\footnotesize{}
EBAC31A08 \citep{FBS+03,SWM+05}} & {\footnotesize --} & {\footnotesize Q9F7M8\_PRB01} & {\footnotesize $\mathrm{C_{2222}H_{3505}N_{607}O_{675}S_{19}}$} & {\footnotesize 447}\tabularnewline
{\footnotesize B} & \emph{\footnotesize Deinococcus radiodurans} & {\footnotesize 25--30 \citep{McD01}} & {\footnotesize ACCA\_DEIRA (Q9RV16) ({*})} & {\footnotesize $\mathrm{C_{1494}H_{2419}N_{453}O_{455}S_{8}}$} & {\footnotesize 316}\tabularnewline
{\footnotesize B} & \emph{\footnotesize Planctomyces maris}{\footnotesize{} \citep{FBS+03}} & {\footnotesize 30 \citep{BS76}} & {\footnotesize A6CDM2\_9PLAN} & {\footnotesize $\mathrm{C_{1611}H_{2578}N_{436}O_{477}S_{13}}$} & {\footnotesize 322}\tabularnewline
{\footnotesize B} & \emph{\footnotesize Actinobacterium}{\footnotesize{} PHSC20C1}\emph{\footnotesize{}
}{\footnotesize \citep{WSP05}} & {\footnotesize $\sim$30 \citep{Lee07}} & {\footnotesize A4AGS7\_9ACTN} & {\footnotesize $\mathrm{C_{3108}H_{4983}N_{889}O_{981}S_{19}}$} & {\footnotesize 675}\tabularnewline
{\footnotesize B} & \emph{\footnotesize $\alpha$-proteobacterium }{\footnotesize (}\emph{\footnotesize C.
crescentus}{\footnotesize )} & {\footnotesize 30} & {\footnotesize ACCA\_CAUCR (Q9A448) ({*})} & {\footnotesize $\mathrm{C_{1517}H_{2440}N_{450}O_{464}S_{11}}$} & {\footnotesize 320}\tabularnewline
{\footnotesize B} & \emph{\footnotesize $\delta$-proteobacterium }{\footnotesize (}\emph{\footnotesize D.
vulgaris}{\footnotesize ) \citep{FBS+03}} & {\footnotesize 37 \citep{LFF+06}} & {\footnotesize A1VC70\_DESVV} & {\footnotesize $\mathrm{C_{3761}H_{5876}N_{1084}O_{1085}S_{29}}$} & {\footnotesize 751}\tabularnewline
{\footnotesize A} & \emph{\footnotesize Methanococcus maripaludis} & {\footnotesize 35--40 \citep{JPG83}} & {\footnotesize A6VIX9\_METM7} & {\footnotesize $\mathrm{C_{2443}H_{3887}N_{659}O_{729}S_{17}}$} & {\footnotesize 494}\tabularnewline
{\footnotesize B} & \emph{\footnotesize Cyanobacterium }{\footnotesize (}\emph{\footnotesize Synechococcus}{\footnotesize )
\citep{ABR+06}} & {\footnotesize 50 \citep{ABR+06}} & {\footnotesize Q2JSS7\_SYNJA ({*})} & {\footnotesize $\mathrm{C_{1608}H_{2590}N_{464}O_{478}S_{9}}$} & {\footnotesize 327}\tabularnewline
{\footnotesize B} & \emph{\footnotesize Chloroflexus aggregans}{\footnotesize{} \citep{FBS+03}} & {\footnotesize 55 \citep{HHS+95}} & {\footnotesize A0GZU2\_9CHLR} & {\footnotesize $\mathrm{C_{2831}H_{4515}N_{821}O_{827}S_{17}}$} & {\footnotesize 590}\tabularnewline
{\footnotesize B} & \emph{\footnotesize Hydrogenobaculum }{\footnotesize sp. Y04AAS1}\emph{\footnotesize{}
}{\footnotesize \citep{SWM+05}} & {\footnotesize 58 \citep{FLB+07}} & {\footnotesize A7WGI1\_9AQUI} & {\footnotesize $\mathrm{C_{2360}H_{3732}N_{622}O_{702}S_{13}}$} & {\footnotesize 472}\tabularnewline
{\footnotesize B} & \emph{\footnotesize Hydrogenobacter thermophilus}{\footnotesize{} \citep{SWM+05}} & {\footnotesize 70--75 \citep{KIK+84}} & {\footnotesize Q05KD0\_HYDTH} & {\footnotesize $\mathrm{C_{2218}H_{3568}N_{612}O_{641}S_{19}}$} & {\footnotesize 444}\tabularnewline
{\footnotesize B} & \emph{\footnotesize Aquifex aeolicus}{\footnotesize{} \citep{SWM+05}} & {\footnotesize 85 \citep{HWH+92}} & {\footnotesize ACCA\_AQUAE (O67260) ({*})} & {\footnotesize $\mathrm{C_{1635}H_{2616}N_{440}O_{457}S_{14}}$} & {\footnotesize 323}\tabularnewline
\hline
\end{tabular}
\end{table}



\section{Code}

Load CHNOSZ.

\begin{wrapsweave}
<<libraryCHNOSZ>>=

library(CHNOSZ)

@
\end{wrapsweave}


Put the names of the model proteins in the variables \texttt{rubisco}
and \texttt{accoaco}:

\begin{wrapsweave}
<<proteins>>=

# ribulose phosphate carboxylase

# in order of increasing growth temperature

rubisco <- c("RBL_BRAJA","A6YF84_9PROT","A1E8R4_9CHLO","A8C9T6_9MYCO","A3EQE1_9BACT","A5CKC7_9CHRO",

  "RBL_SYNJA","Q6JAI0_9RHOD","RBL_METJA","A3DND9_STAMF","A1RZJ5_THEPD","RBL_PYRHO")

rubisco.organisms <- c("a-proteobacterium-R","b-proteobacterium","Bracteacoccus","Mycobacterium",

  "Leptospirillum","Cyanobium","Synechococcus","Cyanidiales",

  "Methanococcus-R","Desulfurococcus","Thermofilum","Pyrococcus")

accoaco <- c("Q9F7M8_PRB01","ACCA_DEIRA","A6CDM2_9PLAN","A4AGS7_9ACTN","ACCA_CAUCR","A1VC70_DESVV",

  "A6VIX9_METM7","Q2JSS7_SYNJA","A0GZU2_9CHLR","A7WGI1_9AQUI","Q05KD0_HYDTH","ACCA_AQUAE")

accoaco.organisms <- c("g-proteobacterium","Deinococcus","Planctomyces","Actinobacterium",

  "a-proteobacterium-A","d-proteobacterium","Methanococcus-A","Synechococcus",

  "Chloroflexus","Hydrogenobaculum","Hydrogenobacter","Aquifex")

@
\end{wrapsweave}


Set up the basis species. The activities of basis species (except
for $\log a_{\mathrm{H_{2}}}$, which is specified in the calculations
below) are set to $\log a_{\mathrm{H_{2}O}}=0$, $\log a_{\mathrm{CO_{2}}}=-3$,
$\log a_{\mathrm{NH_{3}}}=-4$, $\log a_{\mathrm{H_{2}S}}=-7$ and
$\mathrm{pH}=7$.

\begin{wrapsweave}
<<basis>>=

basis(c("CO2","H2O","NH3","H2","H2S","H+"),

  c("aq","liq","aq","aq","aq","aq"),c(-3,0,-4,-6,-7,-7))

@
\end{wrapsweave}


Define a function for $\log a_{\mathrm{H_{2}}}$ as a function of
temperature. This is the same function used in Ref. \citep{DS11}.

\begin{wrapsweave}
<<getlogaH2>>=

get.logaH2 <- function(T) return(-11+T*3/40)

@
\end{wrapsweave}


Write a function, named \texttt{plot.it}, to make the equilibrium
activity diagrams, making use of the \texttt{affinity} and \texttt{diagram}
functions in CHNOSZ. The function contains additional code for coloring
the lines, adding legends to plots, and shading a rectangular area
of the plot. \texttt{plot.it} also returns the result of the \texttt{diagram}
function (the calculated equilibrium activities of the proteins) for
use in further calculations. Instead of printing out the entire function
here, only the argument list is shown.

\begin{wrapsweave}
<<plot.it,echo=FALSE>>=

# function to make various equilibrium activity diagrams

# for the carboxylases

plot.it <- function(vars='H2',lims=c(-15,0),

  do.plot=TRUE,dolegend=NULL,ylim=c(-6,-2),

  rect=NULL,drawme=c("rubisco","accoaco")) {

  proteins <- c('rubisco','accoaco')

  r.p <- get(proteins[1])

  a.p <- get(proteins[2])

  my.p <- list(r.p,a.p)

  species(c(r.p,a.p))

  if(is.list(lims)) myargs <- lims else myargs <- list(lims)

  names(myargs) <- vars

  a <- do.call(affinity,myargs)

  # colors: reds for thermo, blue for meso

  r.cols <- c(rep("blue",6),rep("red",6))

  r.lty <- rep(3,12)

  r.lwd <- rep(2,length(r.p))

  a.cols <- c(rep("blue",6),rep("red",6))

  a.lty <- rep(1,12)

  a.lwd <- rep(2,length(a.p))

  if(!"rubisco" %in% drawme) r.lty <- rep(0,length(r.p))

  if(!"accoaco" %in% drawme) {

    r.lty <- a.lty

    a.lty <- rep(0,length(a.p))

  }

  my.cols <- list(r.cols,a.cols)

  my.lty <- list(r.lty,a.lty)

  my.lwd <- list(r.lwd,a.lwd)

  # setup the plot

  if(do.plot) {

    # leave the top y-axis open if we are plotting as a 

    # function of two variables

    if(length(vars)==1) {

      side <- 1:4

      xline <- 1

    } else {

      side <- c(1,2,4)

      xline <- 2.5

    }     

    thermo.plot.new(xlim=range(myargs[[1]]),ylim=ylim,

      xlab=axis.label(vars[1]),ylab=expression(log~italic(a)),

      yline=2,mar=c(2.5,3,xline,1),side=side)

    if(!is.null(rect)) rect(rect[1],rect[2],rect[3],rect[4],col="grey80",border=NA)

  }

  # calculate equilibrium activities

  # use unit total activity of amino acid residue

  d <- diagram(a,residue=TRUE,col=c(r.cols,a.cols),lty=c(r.lty,a.lty),

    legend.x=NULL,do.plot=do.plot,cex=2,lwd=c(r.lwd,a.lwd),

    mam=FALSE,add=TRUE,logact=0)

  # legends and possibly second y-axis

  if(do.plot) {

    if(!is.null(dolegend)) {

      if(length(dolegend)==2) {

        which.p <- match(dolegend[1],proteins)

        legend('topleft',legend=c(dolegend[1],my.p[[which.p]]),

          lty=c(NA,my.lty[[which.p]]),col=c(NA,my.cols[[which.p]]),

          lwd=c(NA,my.lwd[[which.p]]),cex=0.6,bg='white')

      }

      ld <- length(dolegend)

      which.p <- match(dolegend[ld],proteins)

      legend('topright',legend=c(dolegend[ld],my.p[[which.p]]),

        lty=c(NA,my.lty[[which.p]]),col=c(NA,my.cols[[which.p]]),

        lwd=c(NA,my.lwd[[which.p]]),cex=0.6,bg='white')

    }

    # add a second y-axis

    #if(length(vars)==2) secondaxis(vars[2],range(lims[[2]]))

  }

  return(invisible(list(a=a,d=d)))

}

@
\end{wrapsweave}


\begin{wrapsweave}
<<plot.it.args>>=

print(args(plot.it))

@
\end{wrapsweave}


Define the limits of temperature and $\log a_{\mathrm{H_{2}}}$ that
we will use on the plots.

\begin{wrapsweave}
<<limits>>=

Tlim <- c(0,140)

H2lim <- c(-10,-2)

Tval <- seq(Tlim[1],Tlim[2],length.out=128)

H2val <- seq(H2lim[1],H2lim[2],length.out=128)

@
\end{wrapsweave}


Identify the low- and high-temperature model organisms.

\begin{wrapsweave}
<<itemp>>=

ilowT <- c(1:6,13:18)

ihighT <- c(7:12,19:24)

@
\end{wrapsweave}


Define the chemical activity cutoff for counting chemical species
richness.

\begin{wrapsweave}
<<loga.ref>>=

loga.ref <- -5

@
\end{wrapsweave}



\section{Results}


\subsection{\label{sub:1}Changing a single variable}

Fig. \ref{sub:1}a shows the metastable equilibrium activities of
the proteins as a function of temperature at $\log a_{\mathrm{H_{2}}}=-6$,
and Fig. \ref{sub:1}b shows the metastable equilibrium activities
of the proteins as a function of $\log a_{\mathrm{H_{2}}}$ at $T=25$
$^{\circ}$C. The calculations were carried out for unit total activity
of amino acid residues.

%% change figure width
\setkeys{Gin}{width=1.0\textwidth}

\begin{wrapsweave}
<<carboxylase1,fig=T,results=hide>>=

# Figure 1: relative abundances and richness f(T,logaH2)

par(mfrow=c(1,2))

par(cex=1,lwd=1,mar=c(2.5,3,1,1))

# Figure 1a: relative abundances as a function of temperature

Tout <- plot.it('T',Tlim,dolegend='accoaco',rect=c(Tlim[1],-6,Tlim[2],loga.ref))

label.plot('a',xfrac=0.1)

# Figure 1b: relative abundances as a function of logaH2

H2out <- plot.it('H2',H2lim,dolegend='rubisco',rect=c(H2lim[1],-6,H2lim[2],loga.ref))

label.plot('b',xfrac=0.1)

@
\end{wrapsweave}


The figures each have all 24 proteins. The legend is split between
the two panes, with AcCoACo listed on the left (solid lines) and RuBisCO
on the right (dotted lines). The colors of the lines indicate the
growth temperatures of the organisms taken from Table \ref{tab:carboxylase}.
The two figures are approximately mirror images of each other; increasing
$\log a_{\mathrm{H_{2}}}$ at constant temperature (right) has a similar
effect as decreasing temperature at constant $\log a_{\mathrm{H_{2}}}$
(left). According to the figures, either of these changes tends to
stabilize the proteins from the organisms adapted to growth at high
temperature; that is, there are more red lines higher on the diagram
at either low temperature or high $\log a_{\mathrm{H_{2}}}$. 


\subsection{\label{sub:2}Counting species}

If we count the numbers of red and blue lines that are above the shaded
area in Fig. \ref{sub:1}, we get the diagrams shown in Fig. \ref{sub:2}.
These figures, constructed using the \texttt{revisit} function in
CHNOSZ, show predicted richness, i.e. counts of chemical species whose
activities are above a certain value ($-3.4$ in this case). The red
line, which indicates the sum of proteins from organisms adapted to
growth at high temperature, moves to higher values at low temperature
and/or high $\log a_{\mathrm{H_{2}}}$.

\begin{wrapsweave}
<<carboxylase2,fig=T,results=hide>>=

# Figure 2: richness plot as a function of temperature

par(mfrow=c(1,2))

par(cex=1,lwd=1,mar=c(2.5,3,1,1))

revisit(Tout$d,"richness",loga.ref=loga.ref,yline=2,plot.ext=FALSE)

revisit(Tout$d,"richness",loga.ref=loga.ref,ispecies=ilowT,col="blue",add=TRUE,plot.ext=FALSE)

revisit(Tout$d,"richness",loga.ref=loga.ref,ispecies=ihighT,col="red",add=TRUE,plot.ext=FALSE)

label.plot("a",xfrac=0.1)

# Figure 2b: richness plot as a function of logaH2

revisit(H2out$d,"richness",loga.ref=loga.ref,yline=2,plot.ext=FALSE)

revisit(H2out$d,"richness",loga.ref=loga.ref,ispecies=ilowT,col="blue",add=TRUE,plot.ext=FALSE)

revisit(H2out$d,"richness",loga.ref=loga.ref,ispecies=ihighT,col="red",add=TRUE,plot.ext=FALSE)

label.plot("b",xfrac=0.1)

@
\end{wrapsweave}


If you're like me, you wonder why low temperatures apparently favor
formation of the proteins from no-so-hot organisms. Conversely, high
$\log a_{\mathrm{H_{2}}}$ apparently favors the hot organisms, which
might make sense if hot environments are also relatively reducing.


\subsection{\label{sub:3}As a function of two variables}

Let's look at what happens when we vary temperature and $\log a_{\mathrm{H_{2}}}$
independently. The following diagram is an equilibrium predominance
diagram with all 24 proteins as candidates; only a few most stable
proteins appear (they correspond to the highest lines in Fig. \ref{sub:1}).

%% change figure width
\setkeys{Gin}{width=0.6\textwidth}

\begin{wrapsweave}
<<carboxylase3,fig=T,results=hide>>=

# Figure 3: logaH2-T protein stability diagram 

par(cex=1.5,lwd=1.5)

# first get the logaH2 values

logaH2.line <- get.logaH2(Tval)

# calculate affinities; proteins were loaded previously

a <- affinity(T=Tlim,H2=H2lim)

# plot the protein stabilities

col <- rep(c(rep("blue",6),rep("red",6)),2)

diagram(a,residue=TRUE,col.names=col,color=NULL)

# as well as water reduction line, taking fH2=Psat

s <- subcrt(c("H2","H2"),c("gas","aq"),c(-1,1),T=Tval)

logaH2 <- s$out$logK + log10(s$out$P)

lines(Tval,logaH2,lwd=4,lty=2,col="forestgreen")

# and show our logaH2 function

lines(Tval,logaH2.line,lwd=4,lty=3,col="forestgreen")

legend("bottomright",legend=c("H2O reduction","logaH2 = -11 + T*3/40"),

  lty=c(2,3),lwd=3,col="forestgreen")

@
\end{wrapsweave}


The color code used for the protein names is the same as above: red
for organisms from higher temperatures (the last six organisms in
each part of Table \ref{tab:carboxylase}) and blue for the others.
The dashed green line indicates the reduction stability limit of $\mathrm{H_{2}O}$,
i.e. the value of $\log a_{\mathrm{H_{2}}}$ taken from\[
\mathrm{H_{2}}_{\left(g\right)}\rightleftarrows\mathrm{H_{2}}_{\left(aq\right)}\,,\]
using $\log f_{\mathrm{H_{2}}_{\left(g\right)}}=1$ below 100 $^{\circ}$C
and $\log f_{\mathrm{H_{2}}_{\left(g\right)}}=P_{\mathrm{sat}}$,
or pressure of vapor-liquid equilibrium of $\mathrm{H_{2}O}$, at
or above 100 $^{\circ}$C. The dotted green line is a hypothetical
trajectory of $\log a_{\mathrm{H_{2}}}$ given by

\[
\log a_{\mathrm{H_{2}}}=-11+T*3/40\,.\]
This equation is simply a proposal for a line cutting across the predominance
diagram that lands in the low-temperature-protein fields at low temperatures
and the high-temperature-protein fields at high temperatures.


\subsection{\label{sub:4}Along the temperature-oxidation trajectory}

Now let's make a chemical activity diagram showing the equilibrium
activities of the proteins when temperature and $\log a_{\mathrm{H_{2}}}$
are varied simultaneously according to the equation above. This ends
up using the provision in \texttt{affinity} for specifying arbitrary
increments of variables.

%% change figure width
\setkeys{Gin}{width=0.9\textwidth}

\begin{wrapsweave}
<<carboxylase4,fig=T,results=hide>>=

    # Figure 4: relative abundances with logaH2 as linear f(T)

    par(mfrow=c(2,1))

    # Figure 3a: relative abundances with logaH2 as linear f(T)

    mylogaH2 <- get.logaH2(Tval)

    out <- plot.it(c("T","H2"),list(Tval,mylogaH2),rect=c(Tlim[1],-6,Tlim[2],loga.ref))

    label.plot("a",xfrac=0.1)

    # Figure 3b: richness with linear logaH2

    revisit(out$d,"richness",loga.ref=loga.ref,yline=2,side=c(1,2,4),plot.ext=FALSE)

    revisit(out$d,"richness",loga.ref=loga.ref,ispecies=ilowT,

      col="blue",add=TRUE,plot.ext=FALSE)

    revisit(out$d,"richness",loga.ref=loga.ref,ispecies=ihighT,

      col="red",add=TRUE,plot.ext=FALSE)

    label.plot("b",xfrac=0.1)

    # to plot a second x-axis  

    usr <- par('usr')

    usr[1:2] <- range(mylogaH2)

    opar <- par(usr=usr)

    thermo.axis(axis.label("H2"),side=3,line=par("mgp")[1],cex=par("cex"))

@
\end{wrapsweave}


Now, we see that higher temperatures coupled with higher values of
$\log a_{\mathrm{H_{2}}}$ stabilize the proteins from organisms adapted
to growth at higher temperatures.


\subsection{\label{sub:5}Toward species abundance distributions}

Another way of portraying the chemical activities shown in Fig. \ref{sub:4}
above is with a rank-activity diagram. 

\begin{wrapsweave}
<<plot.rank,echo=FALSE>>=

plot.rank <- function(T=25,ntop=5,pretext=NULL,lcex=1) {

  proteins <- c('rubisco','accoaco')

  organisms <- c(rubisco.organisms,accoaco.organisms)

  # new scheme 20090611: red for hot, blue for cold

  # open for rubisco, filled for accoaco

  col <- rep(c(rep("blue",6),rep("red",6)),2)

  pch <- c(rep(c(0:2,5:7),2),rep(c(15:20),2))

  H2 <- get.logaH2(T)

  if(length(T)==1) {

    basis("H2",H2)

    a <- affinity(T=T)

  } else a <- affinity(T=T,H2=H2)

  d <- diagram(a,residue=TRUE,legend.x=NULL,ylim=c(-6,-2),do.plot=FALSE)

  # for each point make a rank plot

  rank <- 1:length(d$logact)

  # put some lead-in and lead-out frames

  res <- length(T)

  ido <- c(rep(1,15),1:res,rep(res,20))

  if(res==1) ido <- 1

  for(i in ido) {

    loga <- numeric()

    for(j in 1:length(d$logact)) loga <- c(loga,d$logact[[j]][i])

    if(i > 4) myrank4 <- myrank3

    if(i > 3) myrank3 <- myrank2

    if(i > 2) myrank2 <- myrank1

    if(i > 1) myrank1 <- myrank

    order <- order(loga,decreasing=TRUE)

    myrank <- rank(loga)

    cex <- rep(1.2,24)

    # show changed by increasing point size

    # any points that changed on the step before the step 

    # before the step before?

    if(i > 4) {

      ichanged <- myrank3 != myrank4

      cex[ichanged[order]] <- cex[ichanged[order]] + 0.1

    }

    # any points that changed on the step before the step before?

    if(i > 3) {

      ichanged <- myrank2 != myrank3

      cex[ichanged[order]] <- cex[ichanged[order]] + 0.2

    }

   # any points that changed on the step before?

    if(i > 2) {

      ichanged <- myrank1 != myrank2

      cex[ichanged[order]] <- cex[ichanged[order]] + 0.3

    }

    # any points that changed on this step?

    if(i > 1) {

      ichanged <- myrank != myrank1

      cex[ichanged[order]] <- cex[ichanged[order]] + 0.4

    }

    plot(rank,loga[order],col=col[order],pch=pch[order],

      ylab=expression(log~italic(a)),cex=cex,cex.main=1,cex.lab=1,cex.axis=1)

    myT <- format(round(T,1))[i]

    myH2 <- format(round(H2,2))[i]

    if(!is.null(pretext)) title(main=substitute(list(Z~X~degree*C,

      log*italic(a)[paste(H2)]==Y),list(X=myT,Y=myH2,Z=pretext)))

    else title(main=substitute(list(X~degree*C, log*italic(a)[paste(H2)]==Y),

      list(X=myT,Y=myH2)))

    legend("topright",legend=c(paste("top",ntop),organisms[order[1:ntop]]),

      pch=c(NA,pch[order[1:ntop]]),col=c(NA,col[order[1:ntop]]),

      pt.cex=c(NA,cex[1:ntop]),cex=lcex)

    order <- order(loga)

    legend("bottomleft",legend=c(paste("low",ntop),organisms[order[ntop:1]]),

      pch=c(NA,pch[order[ntop:1]]),col=c(NA,col[order[ntop:1]]),

      pt.cex=c(NA,cex[24:(24-ntop+1)]),cex=lcex)

  }

}



@
\end{wrapsweave}


%% change figure width
\setkeys{Gin}{width=0.5\textwidth}

\begin{wrapsweave}
<<carboxylase5,fig=T,results=hide>>=

plot.rank(T=25)

@
\end{wrapsweave}


The open symbols represent RuBisCo and the closed symbols AcCoACo.
Note that the proteins with the highest activity in this figure are
the same as those with the highest activities at $T=25$ $^{\circ}$C
in Fig. \ref{sub:4}. Next up are the 100 $^{\circ}$C results.

\begin{wrapsweave}
<<carboxylase6,fig=T,results=hide>>=

plot.rank(T=100)

@
\end{wrapsweave}


Going from low to high temperature (along with the associated changes
in $\log a_{\mathrm{H_{2}}}$) there is clearly a major shift in the
structure of the metastable assemblage of proteins. Not only do the
most dominant (in metastable equilibrium) proteins change, but the
overall shape goes from something that resembles a lognormal distribution
to one that is more even.

The plotting was actually done using a function defined here called
\texttt{plot.rank}. The function is too long to print out here, so
check it out in the Sweave source for this vignette. The function
might seem long and intimidating, but a lot of it is there to add
the title and legends and to deal with sequences of plots (i.e., increasing
the point sizes for proteins that changed rank in the previous few
frames). For example, the following bit of code produces a bunch of
png files along a temperature/$\log a_{\mathrm{H_{2}}}$ profile.

\begin{wrapsweave}
<<makepng,eval=FALSE>>=

png(filename="Rplot%03d.png",width=420,height=320)

plot.rank(T=25:125,lcex=0.8)

dev.off()

@
\end{wrapsweave}


The png files can be converted to an animated GIF using the ImageMagick
(\url{http://imagemagick.org}) \texttt{convert} program (e.g., \texttt{convert
-loop 0 -delay 15 Rplot{*}.png movie.gif}). A movie produced using
this method (\texttt{carboxylase.gif}) is available online at the
CHNOSZ website (\url{http://chnosz.net/vignettes}).


\section{Conclusions}

The calculations described above summarize a simple chemical thermodynamic
model that uses protein reactions to study the adaptation of organisms
to different growth temperatures. Using the parameters in the model,
changing temperature alone produces results that do not align with
the actual growth temperatures of organisms. When a temperature-dependent
value of oxidation potential (expressed as activity of dissolved hydrogen)
is imposed on the system, the predicted stabilities of the proteins
fit better with the growth temperatures of the organisms.

\bibliographystyle{plainnat}
\bibliography{vig}

\end{document}
