%% LyX 1.6.7 created this file.  For more info, see http://www.lyx.org/.
%% Do not edit unless you really know what you are doing.
\documentclass[10pt,english,slidestop,compress,mathserif]{beamer}
\usepackage{mathpazo}
\usepackage[T1]{fontenc}
\usepackage[latin9]{inputenc}
\setcounter{tocdepth}{1}
\usepackage{amsmath}
\usepackage{amssymb}
\usepackage[authoryear]{natbib}

\makeatletter

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%% LyX specific LaTeX commands.
%% Because html converters don't know tabularnewline
\providecommand{\tabularnewline}{\\}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%% Textclass specific LaTeX commands.
 % fix for LyX layout using \DeclareLaTeXClass[beamer, Sweave.sty]
 % (fails with Undefined control sequence \makebeamertitle)
 \newcommand\makebeamertitle{\frame{\maketitle}}%
 \AtBeginDocument{
   \let\origtableofcontents=\tableofcontents
   \def\tableofcontents{\@ifnextchar[{\origtableofcontents}
 {\gobbletableofcontents}}
   \def\gobbletableofcontents#1{\origtableofcontents}
 }
 \makeatletter
 \long\def\lyxframe#1{\@lyxframe#1\@lyxframestop}%
 \def\@lyxframe{\@ifnextchar<{\@@lyxframe}{\@@lyxframe<*>}}%
 \def\@@lyxframe<#1>{\@ifnextchar[{\@@@lyxframe<#1>}{\@@@lyxframe<#1>[]}}
 \def\@@@lyxframe<#1>[{\@ifnextchar<{\@@@@@lyxframe<#1>[}{\@@@@lyxframe<#1>[<*>][}}
 \def\@@@@@lyxframe<#1>[#2]{\@ifnextchar[{\@@@@lyxframe<#1>[#2]}{\@@@@lyxframe<#1>[#2][]}}
 \long\def\@@@@lyxframe<#1>[#2][#3]#4\@lyxframestop#5\lyxframeend{%
   \frame<#1>[#2][#3]{\frametitle{#4}#5}}
 \makeatother
 \def\lyxframeend{} % In case there is a superfluous frame end
 \makeatletter
 \long\def\lyxagainframe#1{\@lyxagainframe#1\@lyxframestop}%
 \def\@lyxagainframe{\@ifnextchar<{\@@lyxagainframe}{\@@lyxagainframe<*>}}%
 \def\@@lyxagainframe<#1>{\@ifnextchar[{\@@@lyxagainframe<#1>}{\@@@lyxagainframe<#1>[]}}
 \long\def\@@@lyxagainframe<#1>[#2]#3\@lyxframestop{%
   \againframe<#1>[#2]{#3}}
 \makeatother

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%% User specified LaTeX commands.
%\VignetteIndexEntry{2. A different kind of stability: Energy of protein formation}
% beamer stuff
%\usepackage{beamerthemeshadow}
\usetheme{Frankfurt}
%removes bottom navigation bars
\setbeamertemplate{footline}[page number]{}
%removes navigation symbols
\setbeamertemplate{navigation symbols}{}
% for some reason doesn't work even with
% beamer option mathserif
%\usepackage{eulervm}

% for scrap sweave environment in LyX
\newenvironment{wrapsweave}{\par}{\par}

% command for caret character in math mode
\newcommand\caret{\mathbin{\char`\^}}

% alternate color scheme 
\newcommand\colorthemeflu{
\setbeamercolor{normal text}{bg=black!30}
\setbeamercolor{item projected}{use=item,fg=black,bg=green}
}

% reset to original colors
\newcommand\colorthemereset{
\setbeamercolor{normal text}{bg=white}
\setbeamercolor{item projected}{parent=item,use=item,fg=white,bg=item.fg}
}

\makeatother

\usepackage{babel}

\begin{document}

\title{A Different Kind Of Stability}


\subtitle{Energy of Protein Formation}


\author{Jeffrey M. Dick}


\institute{School of Earth and Space Exploration\\
Arizona State University}

\makebeamertitle
\begin{wrapsweave}
<<libraryCHNOSZ,echo=F,results=hide>>=

### loading CHNOSZ

library(CHNOSZ)

# end Sweave chunk

@
\end{wrapsweave}


\begin{wrapsweave}
<<setup_four,echo=FALSE,results=hide>>=

prot <- c("LYSC","CYC","RNAS1","MYG")

org <- c("CHICK","BOVIN","BOVIN","PHYCA")

proteins <- po <- paste(prot,org,sep="_")

basis("CHNOS+")

species(po)

pl <- protein.length(po)

loga.prot <- log10(1/pl)

species(1:4,loga.prot)

mymar <- c(3.5,4,0.5,2)

@
\end{wrapsweave}


\begin{wrapsweave}
<<Gresidue,echo=F,results=hide>>=

### plot amino acid Gibbs energies

thermo.postscript("Gresidue.ps",width=6,height=5)

g <- colnames(thermo$group)

i <- info(g)

G <- info(i)$G

# add protein backbone to sidechain groups to get residues

G <- G[1:20] + G[22]

Gf.aa <- G/1000

i <- i[1:20]

f <- info(i)$formula

M <- element(f)$mass

aa <- aminoacids()

mymar <- c(3.5,4,0.5,2)

par(cex=1.5,mar=mymar,las=1,mgp=c(2.5,1,0))

ihp <- which(aa %in% c("A","G","I","L","M","F","P","W","Y","V"))

col <- rep("blue",length(M))

col[ihp] <- "red"

plot(M,Gf.aa,pch=aa,xlab="Mass, g/mol",ylab="",col=col)

thermo.axis(axis.label("DG0f","k"),side=2,line=3.2)

dev.off()

@
\end{wrapsweave}


\begin{wrapsweave}
<<Gprotein,echo=F,results=hide>>=

### plot protein Gibbs energies

thermo.postscript("Gprotein.ps",width=6,height=5)

ip <- info(proteins)

Gf.prot <- info(ip)$G/1000

# highest values of unfolding Gibbs energies reported by Pfeil, 1998

# (Protein Stability and Folding, Springer) convert kJ/mol to kcal/mol

Gunfold <- convert(c(73.6,64.4,79.1,62.8),"cal")

Gfold <- -Gunfold

par(cex=1.5,mar=mymar,las=1,mgp=c(2,0.7,0))

height <- matrix(c(Gf.prot,Gfold),ncol=4,byrow=TRUE)

barplot(height,ylim=c(pretty(Gf.prot)[1],0),ylab=NULL,names=prot,col=c("grey","green4"),border=NA)

thermo.axis(axis.label("DG0f","k"),side=2,line=3.2)

dev.off()

@
\end{wrapsweave}


\begin{wrapsweave}
<<Zprotein_ionized,echo=F,results=hide>>=

### plot charges of CYC_BOVIN

thermo.postscript("Zprotein_ionized.ps",width=6,height=5)

par(cex=1.5,mar=mymar,las=1,mgp=c(2,0.7,0))

  # add the ionizable groups

  ionize()

  # get the affinities and charges (along equal temperature increments)

  T <- c(25,150,6); pH <- c(0,14,50)

  x <- affinity(pH=pH,T=T)

  z <- ionize(x$values,x$values)

  # plot charges at the temperatures we're interested in

  plot(z[[2]][,1],x=seq(0,14,length.out=50),type="l",xlab="pH",

    ylab="net charge (Z)")  # 25  deg C

  lines(z[[2]][,4],x=seq(0,14,length.out=50),col="red")    # 100 deg C

  lines(z[[2]][,6],x=seq(0,14,length.out=50),col="orange") # 150 deg C

  text(x=c(12,6.8),y=c(-15,-17),labels=paste("T=",c(25,150),sep=""))

  ionize(FALSE)

dev.off()

@
\end{wrapsweave}


\begin{wrapsweave}
<<Gprotein_ionized,echo=F,results=hide>>=

### plot Gibbs energies of ionized CYC_BOVIN

thermo.postscript("Gprotein_ionized.ps",width=6,height=5)

par(cex=1.5,mar=mymar,las=1,mgp=c(2,0.7,0))

  # add the ionizable groups

  ionize()

  # get the affinities and charges (along equal temperature increments)

  T <- c(25,150,6); pH <- c(0,14,50)

  a <- affinity(pH=pH,T=T)

  g <- affinity(pH=pH,T=T,property="G.species")

  myg <- ionize(a$values,g$values)

  # plot energies at the temperatures we're interested in

  plot(myg[[2]][,1]/1000000,x=seq(0,14,length.out=50),type="l",xlab="pH",

    ylab="",ylim=c(-4.8,-3.2))  # 25  deg C

  # add an axis label

  mtext("Gibbs energy (Mcal/mol)",2,line=2.5,cex=par("cex"),las=0)

  lines(myg[[2]][,4]/1000000,x=seq(0,14,length.out=50),col="red")    # 100 deg C

  lines(myg[[2]][,6]/1000000,x=seq(0,14,length.out=50),col="orange") # 150 deg C

  text(x=c(2,2,2),y=c(-4.6,-4.2,-3.9),labels=paste(c(150,100,25)," degC",sep=""))

  ionize(FALSE)

dev.off()

@
\end{wrapsweave}


\begin{wrapsweave}
<<table,echo=FALSE,eval=FALSE>>=

# get values for table in presentation

# properties of formation reactions of residues

prot <- c("LYSC","CYC","RNAS1","MYG")

org <- c("CHICK","BOVIN","BOVIN","PHYCA")

po <- paste(prot,org,sep="_")

basis("CHNOS+")

#basis("CHNOS")

species(po)

pl <- protein.length(po)

# equal-activity reference state

# (unit activity of residues)

loga.prot <- log10(1/pl)

species(1:4,loga.prot)

A.in <- affinity()

# we are dealing with the ionized proteins

ionize()

A <- affinity()

# the standard state properties of reactions

logK <- affinity(property="logK")

logK <- as.numeric(ionize(A$values,logK$values))

# the activity products

logQ.act <- affinity(property="logQ")

logQ.act <- as.numeric(ionize(A$values,logQ.act$values))

# done with equal-activity reference state

  # first get to per residue

A.act <- as.numeric(A.in$values)[1:4]/pl

  # then account for the protein-residue activity changes

A.act <- -log10(pl) - (pl-1)/pl*loga.prot + A.act

logK <- logK[1:4]/pl

logQ.act <- logQ.act[1:4]/pl

logQ.act <- log10(pl) + (pl-1)/pl*loga.prot + logQ.act

ionize(FALSE)

# go to equal-affinity reference state

loga.prot <- as.numeric(diagram(A.in,do.plot=FALSE)$logact)

loga.residue <- log10(pl*10^loga.prot)

species(1:4,loga.prot)

A.aff <- affinity()

# again deal with the ionized proteins

ionize()

A <- affinity()

logQ.aff <- affinity(property="logQ")

logQ.aff <- as.numeric(ionize(A$values,logQ.aff$values))

# done with equal-affinity reference state

A.aff <- as.numeric(A.aff$values)[1:4]/pl

A.aff <- -log10(pl) - (pl-1)/pl*loga.prot + A.aff

logQ.aff <- logQ.aff[1:4]/pl

logQ.aff <- log10(pl) + (pl-1)/pl*loga.prot + logQ.aff

ionize(FALSE)

# summarize the results

out <- data.frame(logK=logK,logQ.act=logQ.act,A.act=A.act,loga.prot=loga.prot,loga.residue=loga.residue,logQ.aff=logQ.aff,A.aff=A.aff)

rownames(out) <- po

@
\end{wrapsweave}


\begin{wrapsweave}
<<80O2_four,echo=FALSE,results=hide>>=

# just at logfO2 = -80

thermo.postscript("80O2_four.ps",width=6,height=5)

par(cex=1.5)

a <- affinity()

d <- diagram(a,do.plot=FALSE)

thermo.plot.new(xlim=c(-85,-65),ylim=c(-8,0),xlab=axis.label("O2"),ylab=as.expression(quote(log~italic(a))),mar=mymar,yline=par("mgp")[1]+1)

myloga <- as.numeric(d$logact)

points(rep(-80,4),myloga,pch=19)

# put some space around the text labels

text(rep(-79.5,3),myloga[c(1,2,4)],prot[c(1,2,4)],adj=0)

text(rep(-80.5,1),myloga[3],prot[3],adj=1)

dev.off()

@
\end{wrapsweave}


\begin{wrapsweave}
<<O2_four,echo=FALSE,results=hide>>=

thermo.postscript("O2_four.ps",width=6,height=5)

par(cex=1.5,lwd=1.5)

a <- affinity(O2=c(-85,-65))

d <- diagram(a,ylim=c(-8,0),legend.x=NULL,mar=mymar)

text(-82,-1.3,"MYG")

text(-73.5,-1.2,"CYC")

text(-68,-1.2,"RNAS1")

text(-67.5,-2.5,"LYSC")

points(rep(-80,4),myloga,pch=19,cex=0.7)

dev.off()

@
\end{wrapsweave}


\begin{wrapsweave}
<<O2_four_unity,echo=FALSE,results=hide>>=

thermo.postscript("O2_four_unity.ps",width=6,height=5)

par(cex=1.5,lwd=1.5)

a <- affinity(O2=c(-85,-65))

d <- diagram(a,ylim=c(-8,0),legend.x=NULL,logact=0,mar=mymar)

text(-82,-1.3,"MYG")

text(-73.5,-1.2,"CYC")

text(-68,-1.2,"RNAS1")

text(-67.5,-2.5,"LYSC")

dev.off()

@
\end{wrapsweave}


\begin{wrapsweave}
<<Eh-pH_four,echo=FALSE,results=hide>>=

thermo.postscript("Eh-pH_four.ps",width=6,height=5)

par(cex=1.5,lwd=1.5)

# we change the basis species to look at Eh instead of logfO2

basis(c("CO2","H2O","NH3","H2S","e-","H+"))

a <- affinity(pH=c(0,14),Eh=c(-0.6,0.6))

diagram(a,names=prot,mar=mymar)

water.lines()

dev.off()

@
\end{wrapsweave}


\begin{wrapsweave}
<<Eh-pH_O2,echo=FALSE,results=hide>>=

thermo.postscript("Eh-pH_O2.ps",width=6,height=5)

par(cex=1.5,lwd=1.5)

diagram(a,names=prot,col="darkgrey",col.names="darkgrey",mar=mymar)

# now overlay the logfO2 contours

basis(c("O2","H+","e-"))

species("H2O")

a <- affinity(pH=c(0,14),Eh=c(-1,1))

diagram(a,what="O2",add=TRUE,cex=1.5,lty=3)

water.lines()

dev.off()

@
\end{wrapsweave}


\begin{wrapsweave}
<<setupEFTu,echo=FALSE,results=hide>>=

# set up for EF-Tu study

# read the amino acid compositions of the proteins

f <- system.file("extdata/EF-Tu.aln",package="CHNOSZ")

# we only use six of the proteins

iip <- c(4:6,1:3)

p <- read.fasta(f,pnff=TRUE)[iip,]

# add the proteins to this session

ip <- add.protein(p)

basis("CHNOS+")

species(paste(p$protein,p$organism,sep="_"))

@
\end{wrapsweave}


\begin{wrapsweave}
<<O2_EFTu,echo=F,results=hide>>=

thermo.postscript("O2_EFTu.ps",width=6,height=5)

a <- affinity(O2=c(-80,-66))

col <- c("blue","red","red","red","red","blue")

lty <- c(1,1,1,2,2,2)

# use a small top margin since the titles will be on the slides

mar <- c(3.5,4,0.5,2)

d <- diagram(a,col=col,names=p$organism,legend.x=NULL,mar=mar,cex=1.5,lwd=1.5,lty=lty)

text(-76.5,-3.4,p$organism[1])

text(-78,-3.05,p$organism[2])

lines(c(-77.5,-76.5),c(-3.09,-3.19))

text(-68,-3.08,p$organism[3])

text(-76,-2.63,p$organism[4])

text(-78,-2.89,p$organism[5])

text(-70,-2.8,p$organism[6])

dev.off()

@
\end{wrapsweave}


\begin{wrapsweave}
<<O2-T_EFTu,echo=F,results=hide>>=

thermo.postscript("O2-T_EFTu.ps",width=6,height=5)

a <- affinity(T=c(0,120),O2=c(-80,-50))

diagram(a,names=p$organism,mar=mar,cex=1.5)

dev.off()

@
\end{wrapsweave}


\begin{wrapsweave}
<<T-pH_EFTu,echo=F,results=hide>>=

thermo.postscript("T-pH_EFTu.ps",width=6,height=5)

basis("O2",-60)

a <- affinity(pH=c(0,14),T=c(20,120))

diagram(a,names=p$organism,mar=mymar,cex=1.5)

dev.off()

@
\end{wrapsweave}


\begin{wrapsweave}
<<O2-pH_EFTu,echo=F,results=hide>>=

thermo.postscript("O2-pH_EFTu.ps",width=6,height=5)

a <- affinity(pH=c(0,14),O2=c(-80,-60),T=25)

diagram(a,names=p$organism,mar=mymar,cex=1.5)

dev.off()

@
\end{wrapsweave}


\begin{wrapsweave}
<<pH_EFTu,echo=F,results=hide>>=

#thermo.postscript("pH_EFTu.ps",width=6,height=5)

#a <- affinity(pH=c(0,14),T=80)

#diagram(a,names=p$organism,mar=mymar,cex=1.5)

#dev.off()

@
\end{wrapsweave}



\lyxframeend{}\lyxframe{Motivation}

%
\begin{minipage}[t][1\totalheight]{0.45\columnwidth}%
\begin{block}<1->{Background}
Protein geochemistry is concerned with the occurrence and variation of proteins in all environments on Earth.
\end{block}%
\end{minipage}\hfill{}%
\begin{minipage}[t][1\totalheight]{0.45\columnwidth}%
\begin{exampleblock}<1>{Example}
Amino acid differences between mitochondrial and nuclear proteins
\end{exampleblock}%
\end{minipage}

%
\begin{minipage}[b][1\totalheight][t]{0.45\columnwidth}%
\begin{block}<2->{Hypothesis}
Molecular evolution is a type of chemical reaction.
\end{block}%
\end{minipage}\hfill{}%
\begin{minipage}[b][1\totalheight][t]{0.45\columnwidth}%
\begin{block}<2->{Outline}
\tableofcontents{}
\end{block}%
\end{minipage}


\lyxframeend{}\section{Protein Formation Reactions}


\lyxframeend{}\lyxframe{Folding Reactions}
\begin{columns}%{}


\column{7cm}
\begin{block}
{}

\includegraphics<1>[width=1\columnwidth]{rxn_unfolded}

\includegraphics<2>[width=1\columnwidth]{rxn_folding}

\end{block}

\column{4cm}
\begin{itemize}
\item <2->Folding as a conformational process
\item <2->Stability referenced to unfolded protein
\item <2->Cellular/laboratory timescales
\end{itemize}
\end{columns}%{}



\lyxframeend{}\lyxframe{Formation Reactions}
\begin{columns}%{}


\column{7cm}
\begin{block}
{}

\includegraphics<1>[width=1\columnwidth]{rxn_unfolded}

\includegraphics<2>[width=1\columnwidth]{rxn_formulas}

\includegraphics<3>[width=1\columnwidth]{rxn_formation}

\end{block}

\column{4cm}
\begin{itemize}
\item <2->Formation as a chemical process
\item \only<2>{$\Delta G_{f}^{\circ}$: Standard Gibbs energy of formation
from the elements}\only<3->{Stability referenced to inorganic species}
\item <3->$\Delta G_{r}^{\circ}$: Standard Gibbs energy of reaction
\item <3->Overall energy change is independent of mechanism
\end{itemize}
\end{columns}%{}



\lyxframeend{}\lyxframe{Residue Equivalents}
\begin{columns}%{}


\column{7cm}
\begin{block}
{}

\includegraphics<1>[width=1\columnwidth]{rxn_residues}

\includegraphics<2>[width=1\columnwidth]{rxn_metastability}

\end{block}

\column{4cm}
\begin{itemize}
\item <1->Reactions normalized by protein length
\item <1->Energetic meaning of reaction coefficients
\item <2->Transformation reaction; cellular to evolutionary timescales\end{itemize}
\begin{exampleblock}
<1->{Environment \& Energy}

Shift to lower $\mathrm{O_{2}}$ potential more strongly decreases
Gibbs energy of formation of CYC than RNAS1.\end{exampleblock}
\end{columns}%{}



\lyxframeend{}\section{Calculating Relative Stabilities}


\lyxframeend{}\lyxframe{Standard Gibbs Energies}
\begin{columns}%{}


\column{6cm}
\begin{block}
{}

\includegraphics<1>[width=1\columnwidth]{Gresidue}

\includegraphics<2>[width=1\columnwidth]{Gprotein}\end{block}
\begin{wrapsweave}

\end{wrapsweave}

\column{5cm}
\begin{itemize}
\item <1->Group additivity of aqueous species properties for amino acid
residues \citep{DLH06}
\item <2->Protein size dependence of standard Gibbs energies

\begin{block}
{}

\begin{tabular}{llr}
LYSC & Lysozyme & 129\tabularnewline
CYC & Cytochrome C & 104\tabularnewline
RNAS1 & Ribonuclease A & 124\tabularnewline
MYG & Myoglobin & 153\tabularnewline
\end{tabular}
\end{block}
\item <2->Gibbs energies of folding \citep{PK74} are 1\% or less of energies
of formation. 
\end{itemize}
\end{columns}%{}



\lyxframeend{}

\colorthemeflu


\lyxframeend{}\lyxframe{<1>[label=plan]Computational Plan}
\begin{columns}%{}


\column{9cm}

\begin{itemize}
  \item<1-> Calculate standard Gibbs energies of proteins from amino acid group additivity and equations of state for aqueous species.
  \item<1-> Proceed without accounting for energy of protein folding.
  \only<1>{
    \begin{itemize}
     \item Per mole of protein, energy of folding is small compared to Gibbs energy of chemical formation reaction.
     \item If all proteins are folded, energy of folding tends to cancel in relative stability calculations.
    \end{itemize}
  }
  \item<2-> Calculate equilibrium activities of proteins.
  \only<2>{
    \begin{itemize}
       \item Start with chemical affinities in equal-activity reference state.
       \item Use reference state transformation to calculate equilibrium activities.
    \end{itemize}
  }
  \only<3>{
    \begin{itemize}
    \item Relative stabilities depend on the species (chemical compositions, standard Gibbs energies).
    \item Relative stabilities depend on the environment (temperature, pressure, activities/fugacities of basis species).
    \end{itemize}
  }
  \item<4-> Investigate temperature, oxidation potential, pH, other chemical potentials.
  \only<4>{
    \begin{itemize}
      \item Are reducing conditions associated with hotter environments?
      \item The system is multidimensional; could also vary the chemical potentials of carbon, nitrogen, sulfur.
    \end{itemize}
  }
  \item<5-> Explore the protein universe using model systems.
  \item<5-> CHNOSZ is the software package used for the preceding calculations.
\end{itemize}

\end{columns}%{}

\lyxframeend{}

\colorthemereset


\lyxframeend{}\lyxframe{Metastable Equilibrium}
\begin{columns}%{}


\column{6cm}
\begin{itemize}
\item <1->Chemical affinity is opposite of Gibbs energy change due to reaction
progress,
\item <2->computed as a function of standard Gibbs energy, reaction stoichiometry
and chemical activities.
\item <4->Maxwell-Boltzmann distribution allows for a transformation between
reference states.
\item <5->When the chemical affinities of the formation reactions are all
equal, the proteins are in metastable equilibrium.
\end{itemize}

\column{5.2cm}
\begin{block}
{}
\begin{itemize}
\item <1->$dG=-SdT+VdP-Ad\xi$
\end{itemize}
\only<1>{\begin{tabular}{cc}
$G$ & Gibbs energy\tabularnewline
$S$ & Entropy\tabularnewline
$T$ & Temperature\tabularnewline
$V$ & Volume\tabularnewline
$P$ & Pressure\tabularnewline
$A$ & Chemical Affinity\tabularnewline
$\xi$ & Reaction Progress\tabularnewline
\end{tabular}}
\begin{itemize}
\item <2->$A=2.303RT\log(K/Q)$
\end{itemize}
\only<2>{



\begin{tabular}{cc}
$A$ & Chemical Affinity\tabularnewline
$K$ & Equilibrium Constant\tabularnewline
$Q$ & Activity Product\tabularnewline
$R$ & Gas Constant\tabularnewline
\end{tabular}}

\only<3>{



\begin{tabular}{cc}
$A$ & \multicolumn{1}{l}{$=-\Delta G_{r}$}\tabularnewline
$K$ & \multicolumn{1}{l}{$=10\caret\left(-\Delta G_{r}^{\circ}/2.303RT\right)$}\tabularnewline
$Q$ & \multicolumn{1}{l}{$=10\caret\sum\nu\log a=\prod a^{\nu}$}\tabularnewline
$\nu$ & Reaction Coefficient\tabularnewline
$a$ & Chemical Activity\tabularnewline
\end{tabular}}
\begin{itemize}
\item <4->${\displaystyle \frac{a}{\sum a}=\frac{e^{A/RT}}{\sum e^{A/RT}}}$
\end{itemize}
\end{block}
\begin{exampleblock}
<4->{Equal-activity reference state}

More stable: higher affinity ($A$)

\end{exampleblock}

\begin{exampleblock}
<5>{Equal-affinity reference state}

More stable: higher activity ($a$)\end{exampleblock}
\end{columns}%{}



\lyxframeend{}

\colorthemeflu


\lyxframeend{}\lyxagainframe{<2>{plan}}


\lyxframeend{}

\colorthemereset


\lyxframeend{}\lyxframe{[label=formation]Formation of Ionized Unfolded CYC\_BOVIN}
\begin{columns}%{}


\column{8cm}
\begin{block}
{}

\only<1>{$4.97\mathrm{CO_{2}}+1.38\mathrm{NH_{3}}+0.04\mathrm{H_{2}S}+1.86\mathrm{H_{2}O}+0.08\mathrm{H^{+}}\rightarrow\frac{1}{104}\mathrm{C_{517}H_{825}N_{143}O_{15}S_{4}^{+0.08}}+5.18\mathrm{O_{2}}$}

\only<2->{$4.97\mathrm{CO_{2}}+1.38\mathrm{NH_{3}}+0.04\mathrm{H_{2}S}+1.86\mathrm{H_{2}O}+0.08\mathrm{H^{+}}\rightarrow\mathrm{C_{4.97}H_{7.93}N_{1.38}O_{0.14}S_{0.04}^{+0.08}}+5.18\mathrm{O_{2}}$

~

}

\only<3>{$\log K=-\Delta G_{r}^{\circ}/2.303RT$

~

$\phantom{\log K}=-393.0$}

\only<4>{$\log Q=\log a_{\mathrm{C_{4.97}H_{7.93}N_{1.38}O_{0.14}S_{0.04}^{+0.08}}}+5.18\log f_{\mathrm{O_{2}}}-4.97\log a_{\mathrm{CO_{2}}}-1.38\log a_{\mathrm{NH_{3}}}-0.04\log a_{\mathrm{H_{2}S}}-1.86\log a_{\mathrm{H_{2}O}}-\log a_{\mathrm{H^{+}}}$

~

$\phantom{\log Q}=-393.4$}

\end{block}

\column{4cm}
\begin{itemize}
\item <1->Ionization of proteins using additivity \hyperlink{ionization<1>}{\beamergotobutton{more info?}}
\item <2->Write per-residue formulas.
\end{itemize}
\only<3>{

~

\begin{tabular}{lr}
$T$ & 25 $^{\circ}$C\tabularnewline
$P$ & 1 bar\tabularnewline
\end{tabular}}

\only<4>{

~

\begin{tabular}{lr}
$\log a_{\mathrm{residue}}$ & $0$\tabularnewline
$\log a_{\mathrm{CO_{2}}}$ & $-3$\tabularnewline
$\log a_{\mathrm{H_{2}O}}$ & $0$\tabularnewline
$\log a_{\mathrm{NH_{3}}}$ & $-4$\tabularnewline
$\log f_{\mathrm{O_{2}}}$ & $-80$\tabularnewline
$\log a_{\mathrm{H_{2}S}}$ & $-7$\tabularnewline
$\mathrm{pH}$ & $7$\tabularnewline
\end{tabular}

}\end{columns}%{}



\lyxframeend{}\lyxframe{Formation Properties}
\begin{columns}%{}


\column{10cm}
\begin{exampleblock}
{Values per Residue}

\only<1>{\begin{tabular}{crrrr}
Protein & $\log K$ & $\log Q$ & $\log a$ & $A/2.303RT$\tabularnewline
LYSC & $-361.6$ & $-357.4$ & $0$ & $-4.17$\tabularnewline
CYC & $-393.0$ & $-393.4$ & $0$ & $0.34$\tabularnewline
RNAS1 & $-352.6$ & $-348.4$ & $0$ & $-4.27$\tabularnewline
MYG & $-407.6$ & $-408.6$ & $0$ & $0.96$\tabularnewline
\end{tabular}}

\only<2>{\begin{tabular}{crrrr}
Protein & $\log K$ & $\log Q$ & $\log a$ & $A/2.303RT$\tabularnewline
LYSC & $-361.6$ & $-362.0$ & $-4.62$ & $0.45$\tabularnewline
CYC & $-393.0$ & $-393.5$ & $-0.10$ & $0.45$\tabularnewline
RNAS1 & $-352.6$ & $-353.1$ & $-4.72$ & $0.45$\tabularnewline
MYG & $-407.6$ & $-408.0$ & $0.51$ & $0.45$\tabularnewline
\end{tabular}}

\only<3->{\begin{tabular}{crrrrr}
Protein & $\log K$ & $\log Q$ & $\log a$ & $A/2.303RT$ & $\log a_{\mathrm{protein}}$\tabularnewline
LYSC & $-361.6$ & $-362.0$ & $-4.62$ & $0.45$ & $-6.73$\tabularnewline
CYC & $-393.0$ & $-393.5$ & $-0.10$ & $0.45$ & $-2.12$\tabularnewline
RNAS1 & $-352.6$ & $-353.1$ & $-4.72$ & $0.45$ & $-6.81$\tabularnewline
MYG & $-407.6$ & $-408.0$ & $0.51$ & $0.45$ & $-1.68$\tabularnewline
\end{tabular}}

\end{exampleblock}
\begin{itemize}
\item <1->Equal activity reference state: MYG is more stable.
\item <2->Equal affinity reference state: MYG is still more stable!
\item <3->Molality of residue = molality of protein {*} protein length
\item <3->Activity of residue = activity of protein {*} protein length
(assuming ideality)
\end{itemize}
\end{columns}%{}



\lyxframeend{}\lyxframe{[label=activity_four]Equilibrium Activity Diagrams}
\begin{columns}%{}


\column{6cm}
\begin{block}
{}

\includegraphics<1>[width=1\columnwidth]{80O2_four}

\includegraphics<2>[width=1\columnwidth]{O2_four}

\end{block}

\column{5cm}
\begin{itemize}
\item Metastable equilibrium activities \only<2->{as a function of $\log f_{\mathrm{O_{2}}}$}
with total activity of residues = 4
\item \only<1>{We are using $T=25$ $^{\circ}$C and pH = 7}\only<2>{$\log f_{\mathrm{O_{2}}}$
can be converted to other measurements of oxidation-reduction potential.
\hyperlink{logfO2<1>}{\beamergotobutton{how?}}}\end{itemize}
\begin{exampleblock}
{It Shows ...}

\only<1>{MYG is relatively most stable at $\log f_{\mathrm{O_{2}}}=-80$.}\only<2>{Relative
stability is sensitive to oxidation potential.}\end{exampleblock}
\end{columns}%{}



\lyxframeend{}

\colorthemeflu


\lyxframeend{}\lyxagainframe{<3>{plan}}


\lyxframeend{}

\colorthemereset


\lyxframeend{}\section{Natural Experiment}


\lyxframeend{}\lyxframe{Elongation Factor Tu}
\begin{columns}%{}


\column{6cm}
\begin{block}
{}

\begin{center}
\includegraphics<1>[width=1\columnwidth]{1TUI_small}
\par\end{center}

\begin{center}
\includegraphics<2->[width=0.8\columnwidth]{1TUI_tree}
\par\end{center}

\end{block}

\column{5cm}
\begin{itemize}
\item <1->EF-Tu\only<2->{'s} from \emph{Escherichia coli}\only<2->{,
\emph{Thermotoga maritima}, \emph{Thermus thermophilus}}
\item <2->and reconstructed by maximum likelihood (\textbf{ML}) \textbf{stem}
of bacterial tree, stem of \textbf{meso}philic bacteria, and \textbf{Alt}ernative
tree topology \citep{GTB+03}
\item <2->This tree built using parsimony (PHYLIP software), 394 aligned
amino acids.
\end{itemize}
\end{columns}%{}



\lyxframeend{}\lyxframe{Equilibrium Activity Diagrams}
\begin{columns}%{}


\column{6cm}
\begin{block}
{}

\includegraphics<1>[width=1\columnwidth]{O2_EFTu}

\includegraphics<2>[width=1\columnwidth]{O2-T_EFTu}\includegraphics<3>[width=1\columnwidth]{T-pH_EFTu}

\includegraphics<4>[width=1\columnwidth]{O2-pH_EFTu}%\includegraphics<4>[width=1\columnwidth]{pH_EFTu}

\end{block}

\column{5cm}
\begin{itemize}
\item \only<1>{Equilibrium activity diagram}\only<2->{Equilibrium predominance
diagram}
\item \only<1>{In most reactions, proteins from higher temperature favored
by lower $\log f_{\mathrm{O_{2}}_{\left(g\right)}}$.}\only<2>{At
constant $f_{\mathrm{O_{2}}_{\left(g\right)}}$, increasing $T$ tends
to favor {}``ML-stem''.}\only<3>{$\log f_{\mathrm{O_{2}}}=-60$}\only<4>{$T=25$
$^{\circ}$C}\end{itemize}
\begin{exampleblock}
<4->{Multidimensionality}

What about $\log a_{\mathrm{H_{2}O}}$, $\log a_{\mathrm{CO_{2}}}$,
etc.?\end{exampleblock}
\end{columns}%{}



\lyxframeend{}

\colorthemeflu


\lyxframeend{}\lyxagainframe{<4-5>{plan}}


\lyxframeend{}

\colorthemereset


\lyxframeend{}

\appendix

\lyxframeend{}\lyxframe{References}

\def\newblock{\hskip .11em plus .33em minus .07em}

%\renewcommand\refname{References}

\bibliographystyle{plainnat}
\bibliography{vig}



\lyxframeend{}\section{Protein Ionization}


\lyxframeend{}\lyxframe{[label=ionization]Protein Ionization (CYC\_BOVIN)}
\begin{columns}%{}


\column{6cm}
\begin{block}
{}

\includegraphics<1>[width=1\columnwidth]{Zprotein_ionized}

\includegraphics<2>[width=1\columnwidth]{Gprotein_ionized}

\end{block}

\column{5cm}
\begin{itemize}
\item Net charges computed additively \citep{DLH06} using temperature-dependent
sidechain p\emph{K}a values
\item <2->Also affects standard Gibbs energies of the ionized proteins
\end{itemize}
\only<2>{\hyperlink{formation<1>}{\beamerreturnbutton{return}}}\end{columns}%{}



\lyxframeend{}\section{Oxygen Fugacity}


\lyxframeend{}\lyxframe{[label=logfO2]Oxygen Fugacity}
\begin{columns}%{}


\column{6cm}
\begin{block}
{}

\includegraphics<1>[width=1\columnwidth]{Eh-pH_four}

\includegraphics<2>[width=1\columnwidth]{Eh-pH_O2}

\end{block}

\column{5cm}
\begin{itemize}
\item <1->Eh-pH diagram for proteins
\item <1->\only<1>{Dashed lines indicate stability limits of $\mathrm{H_{2}O}$:
$\log f_{\mathrm{O_{2}}}=0$ (upper), $\log f_{\mathrm{O_{2}}}=-83.1$
(lower)}\only<2>{Convert between Eh and $\log f_{\mathrm{O_{2}}}$
using law of mass action for $\mathrm{H_{2}O}\rightleftharpoons\frac{1}{2}\mathrm{O_{2}}+2\mathrm{H^{+}}+2e^{-}$}
\item <2->$\mathrm{Eh}=\frac{RT}{F}\mathrm{pe}=-\frac{RT}{F}\log a_{\mathrm{e^{-}}}$
\end{itemize}
\only<2>{\vfill{}
\hyperlink{activity_four<2>}{\beamerreturnbutton{return}}}\end{columns}%{}



\lyxframeend{}
\end{document}
